Use whitelist of approved CFL takers

feuGeneA opened the pull request:
-   [x] Implement ability to query by `txOrigin` rather than just by `takerAddress`, in finding matching orders with amounts left to fill, at https://github.com/0xProject/0x-coordinator-server/blob/fca20c573cd2614e5408db9a90e1b43fe80636cb/ts/src/handlers.ts#L169
-   [x] Document how the server operator can whitelist CFL takers (set environment variable).
-   [x] Parse (and store) whitelist environment variable on server startup, not upon every approval request.
-   [x] After parsing whitelist, convert all addresses to lowercase, and before checking an address against the whitelist, also convert it to lowercase.
-   [ ] Add a test to act as a CFL taker.

feuGeneA committed:
Add some yarn script conveniences

feuGeneA committed:
Find orders by txOrigin too, IFF it's whitelisted

Whitelist is sourced from environment variable TAKER_WHITELIST, which is
expected to hold a JSON array of Ethereum contract addresses.

abandeali1 commented on ts/src/handlers.ts:
```diff
@@ -548,6 +559,7 @@ export class Handlers {
                 signedTransaction,
                 coordinatorOrders,
                 takerAssetFillAmounts,
+                Handlers._isWhitelistedTakerContract(txOrigin) ? txOrigin : undefined,
```
We want to check if the `takerAddress` is whitelisted here. We're dealing with a single taker contract that we know in advance, but we have no way of knowing if `txOrigin` should be whitelisted.

abandeali1 commented on ts/src/models/transaction_model.ts:
```diff
@@ -44,7 +45,13 @@ export const transactionModel = {
+            let queryString = 'transaction.takerAddress = :takerAddress';
+            let querySubstitutions: { takerAddress: string; txOrigin?: string } = { takerAddress: opts.takerAddress };
+            if (opts.txOrigin !== undefined) {
+                queryString += ' OR transaction.txOrigin = :txOrigin';
```
In the case where the taker contract is whitelisted, we should _only_ query by `txOrigin`.

abandeali1 commented on ts/src/models/transaction_model.ts:
```diff
@@ -44,7 +45,13 @@ export const transactionModel = {
             .leftJoinAndSelect('transaction.takerAssetFillAmounts', 'takerAssetFillAmount')
             .where('order.hash IN (:...orderHashes)', { orderHashes });
         if (opts !== undefined && opts.takerAddress !== undefined) {
```
We actually want `takerAddress` to be undefined here. Basically if `takerAddress !== undefined`, query by `takerAddress`. If `txOrigin !== undefined`, query by `txOrigin`.

We could keep that logic all in this function, but I think the intent might be more clear (and more extensible) if we just broke this up into 2 functions.

feuGeneA replied:
I believe I've changed the query logic to match what you've described (here, and in your other comment).

But, I didn't break apart into 2 functions as you suggested.  Given that the query by `txOrigin` occurs in two slightly different situations, it wasn't clear to me where exactly to break things apart, but if you can advise further, I'm happy to do it.

feuGeneA committed:
Move whitelist to query func; chg query per review

Moved whitelist application from the handlers layer to the query layer,
and changed query logic per review from @abandeali1.

abandeali1 reviewed (approved):
This looks like it should do the job!

feuGeneA commented on ts/src/models/transaction_model.ts:
```diff
@@ -12,6 +12,11 @@ import { utils } from '../utils';
 import { orderModel } from './order_model';
 
+function isWhitelistedTakerContract(address: string): boolean {
+    const whitelist = JSON.parse(process.env.TAKER_WHITELIST || '[]');
```
I updated this PR's checklist to add these two items (repeated below, for your convenience).  Please let me know if there's anything I neglected to capture there.  Thanks for the feedback.

> -   [ ] Parse (and store) whitelist environment variable on server startup, not upon every approval request.
> -   [ ] After parsing whitelist, convert all addresses to lowercase, and before checking an address against the whitelist, also convert it to lowercase.

feuGeneA committed:
Include taker contract whitelist in configs

feuGeneA committed:
eliminate typeorm query logging from test output

feuGeneA committed:
add some CFL tests (incl. some incomplete ones)
